File: frontend/src/api/editFiles.ts
import { apiFetch } from './http'

export async function getEditFiles() {
  return apiFetch<import('../types').EditFileEntry[]>('/edit-files')
}

export async function getEditFileContent(key: string) {
  return apiFetch<{ key: string; content: string }>(`/edit-files/${key}`)
}

export async function saveEditFile(key: string, content: string) {
  return apiFetch(`/edit-files/${key}`, {
    method: 'PUT',
    body: JSON.stringify({ content }),
  })
}

export async function saveEditFilesBatch(files: Record<string, string>, runDoall = false) {
  return apiFetch<{ message: string; detail?: string }>('/edit-files/batch', {
    method: 'POST',
    body: JSON.stringify({ files, run_doall: runDoall }),
  })
}

export interface EditFileTransferNodeResult {
  node_id: number
  node_name: string | null
  status: 'success' | 'failed' | 'skipped'
  transferred_files: string[]
  failed?: Array<{ file: string; error: string }>
  error?: string | null
  doall_output?: string | null
}

export interface EditFileTransferResult {
  success: boolean
  message: string
  source_node_id: number
  source_node_name: string
  files: string[]
  file_keys: string[]
  run_doall: boolean
  nodes_success: number
  nodes_failed: number
  nodes_skipped: number
  total_transferred: number
  per_node: EditFileTransferNodeResult[]
}

export async function transferEditFiles(payload: {
  file_keys: string[]
  target_node_ids?: number[] | null
  all_online?: boolean
  source_node_id?: number | null
  run_doall?: boolean
  content_overrides?: Record<string, string> | null
}) {
  return apiFetch<EditFileTransferResult>('/edit-files/transfer', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

File: backend/app/routers/edit_files.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.auth import require_admin
from app.database import get_db
from app.models import User
from app.schemas import MessageResponse
from app.services.action_log import log_action
from app.services.edit_files_transfer import run_edit_files_transfer
from app.services.file_editor import EDITABLE_FILES, FileEditorService
from app.services.node_manager import get_active_adapter, get_active_node
from app.services.node_sync.config_sync import maybe_replicate_config_files
from app.services.node_sync.groups import require_ha_primary_for_config_ops

router = APIRouter(prefix="/edit-files", tags=["edit-files"])


class FileContentUpdate(BaseModel):
    content: str = ""


class BatchUpdate(BaseModel):
    files: dict[str, str] = {}
    run_doall: bool = True


class TransferRequest(BaseModel):
    file_keys: list[str] = Field(min_length=1)
    target_node_ids: list[int] | None = None
    all_online: bool = False
    source_node_id: int | None = None
    run_doall: bool = False
    content_overrides: dict[str, str] | None = None


def _filename_for_key(file_key: str) -> str:
    fname = EDITABLE_FILES.get(file_key)
    if not fname:
        raise ValueError("Неизвестный файл")
    return fname


@router.get("")
def list_edit_files(_: User = Depends(require_admin)):
    return FileEditorService().list_files()


@router.get("/{file_key}")
def read_edit_file(file_key: str, db: Session = Depends(get_db), _: User = Depends(require_admin)):
    try:
        adapter = get_active_adapter(db)
        content = adapter.read_config_file(_filename_for_key(file_key))
        return {"key": file_key, "content": content}
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.put("/{file_key}")
def save_edit_file(
    file_key: str,
    payload: FileContentUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    require_ha_primary_for_config_ops(db)
    try:
        adapter = get_active_adapter(db)
        adapter.write_config_file(_filename_for_key(file_key), payload.content)
        output = adapter.apply_config_changes()
        from app.services.openvpn_multihome import maybe_ensure_node_openvpn_multihome

        maybe_ensure_node_openvpn_multihome(adapter, get_active_node(db))
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"Файл сохранён, но doall.sh ошибка: {exc}") from exc
    maybe_replicate_config_files(
        db,
        node_id=get_active_node(db).id,
        file_keys=[file_key],
        run_doall=True,
        content_overrides={file_key: payload.content},
    )
    return MessageResponse(message="Файл сохранён и применён", detail=output)


@router.post("/batch", response_model=MessageResponse)
def save_batch(
    payload: BatchUpdate,
    db: Session = Depends(get_db),
    _: User = Depends(require_admin),
):
    require_ha_primary_for_config_ops(db)
    adapter = get_active_adapter(db)
    for key, content in payload.files.items():
        adapter.write_config_file(_filename_for_key(key), content)
    output = None
    if payload.run_doall:
        output = adapter.apply_config_changes()
        from app.services.openvpn_multihome import maybe_ensure_node_openvpn_multihome

        maybe_ensure_node_openvpn_multihome(adapter, get_active_node(db))
    maybe_replicate_config_files(
        db,
        node_id=get_active_node(db).id,
        file_keys=list(payload.files.keys()),
        run_doall=payload.run_doall,
        content_overrides=dict(payload.files),
    )
    return MessageResponse(message="Файлы сохранены", detail=output)


@router.post("/transfer")
def transfer_edit_files(
    payload: TransferRequest,
    db: Session = Depends(get_db),
    user: User = Depends(require_admin),
):
    require_ha_primary_for_config_ops(db)
    try:
        result = run_edit_files_transfer(
            db,
            file_keys=payload.file_keys,
            target_node_ids=payload.target_node_ids,
            all_online=payload.all_online,
            source_node_id=payload.source_node_id,
            run_doall=payload.run_doall,
            content_overrides=payload.content_overrides,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    log_action(
        db,
        action="edit_files_transfer",
        user_id=user.id,
        username=user.username,
        details=(
            f"from={result['source_node_name']};files={','.join(result['files'])};"
            f"success={result['nodes_success']};failed={result['nodes_failed']};"
            f"doall={payload.run_doall}"
        ),
    )
    return result
